cfgrammar/src/lib/span.rs

```diff
@@ -44,72 +44,98 @@ impl Span {
     }
 }
 
-pub struct NewlineToLineColCache<'a> {
-    src: &'a str,
-    high_water: usize,
+pub struct NewlineToLineColCache {
     newlines: Vec<usize>,
+    trailing_bytes: usize,
 }
 
-impl<'a> NewlineToLineColCache<'a> {
-    pub fn new(src: &'a str) -> Self {
-        let newlines: Vec<usize> = vec![0];
+impl Default for NewlineToLineColCache {
+    fn default() -> NewlineToLineColCache {
         NewlineToLineColCache {
-            src,
-            newlines,
-            high_water: 0,
+            newlines: vec![0],
+            trailing_bytes: 0,
         }
     }
+}
 
-    fn offset_to_line_col(&mut self, off: usize) -> (usize, usize) {
-        if self.high_water < off {
-            self.newlines.extend(
-                // assumes off is_char_boundary would return true.
-                self.src[self.high_water..off]
-                    .char_indices()
-                    .filter_map(|c| match c {
-                        (offset, '\n') | (offset, '\r') => Some(offset + 1),
-                        _ => None,
-                    }),
-            );
-
-            self.high_water = off;
-        }
+impl NewlineToLineColCache {
+    /// Feed more input into the cache, calculating newlines data from it.
+    /// The `src` string given is treated as though it is contiguous with previous a call to `feed`.
+    /// subsequent logical lines, and logical bytes reflecting the sum of total of offsets which
+    /// have been fed.
+    pub fn feed(&mut self, src: &str) {
+        let start_pos = self.newlines[self.newlines.len() - 1] + self.trailing_bytes;
+        self.newlines
+            .extend(src.char_indices().filter_map(|c| match c {
+                (offset, '\n') | (offset, '\r') => {
```

Comment:
Will this count a `\n\r` newline (the or `\r\n` or whatever the old DOS standard is) as two newlines?

Reply:
It'll count it as two newlines, I actually wondered about that, and had started working on code to `peek`, but then when i looked it seemed to me that the existing code was doing this treating them separately, thanks for bringing it up I kind of forgot to.

Reply:
Sounds good, I thought I had encountered something, but I don't see it anymore/in the tests.

Reply:
It seemed okay to conservatively special case this to return None, but still zero trailing bytes, even though it'll probably be zero'd by a subsequent `\n` on the next character.

Reply:
Going to resolve with that, but feel free to unresolve if you feel otherwise of course!

Reply:
Sorry about that thanks for bringing it to my attention, I'll just say when something 'I believe this should be resolved by commit id', in the future and let you resolve them if you agree.